lib/shared/src/event/log_event/mod.rs

```diff
@@ -0,0 +1,831 @@
+                `Lookup::is_valid` should catch this during `Lookup` creation, maybe it was not \
+                called?."
+            );
+        };
+
+        for (index, segment) in walker {
+            trace!(%segment, index, "Seeking next segment.");
+            current_pointer = match (segment, current_pointer) {
+                (
+                    SegmentBuf::Field {
+                        name,
+                        requires_quoting: _,
+                    },
+                    Entry::Occupied(entry),
+                ) => match entry.into_mut() {
+                    Value::Map(map) => map.entry(name),
+                    v => return Err(format!("Looking up field on a non-map value: {:?}", v).into()),
+                },
+                (
+                    SegmentBuf::Field {
+                        name,
+                        requires_quoting: _,
+                    },
+                    Entry::Vacant(entry),
+                ) => {
+                    trace!(segment = %name, index, "Met vacant entry.");
+                    return Err(format!(
+                        "Tried to step into `{}` of `{}`, but it did not exist.",
+                        name,
+                        entry.key()
+                    )
+                    .into());
+                }
+                _ => return Err("The entry API cannot yet descend into array indices.".into()),
+            };
+        }
+        trace!(entry = ?current_pointer, "Result.");
+        Ok(current_pointer)
+    }
+
+    /// Returns the entire event as a `Value::Map`.
+    ///
+    /// ```rust
+    /// use shared::{event::*, lookup::*};
+    /// let event = LogEvent::default();
+    /// assert_eq!(event.take(), std::collections::BTreeMap::default());
+    /// ```
+    #[instrument(level = "trace", skip(self))]
+    pub fn take(self) -> BTreeMap<String, Value> {
+        self.fields
+    }
+
+    /// Get a borrow of the contained fields.
+    ///
+    /// ```rust
+    /// use shared::{event::*, lookup::*};
+    /// let mut event = LogEvent::default();
+    /// assert_eq!(event.inner(), &std::collections::BTreeMap::default());
+    /// ```
+    #[instrument(level = "trace", skip(self))]
+    pub fn inner(&self) -> &BTreeMap<String, Value> {
+        &self.fields
+    }
+
+    /// Get a mutable borrow of the contained fields.
+    ///
+    /// ```rust
+    /// use shared::{event::*, lookup::*};
+    /// let mut event = LogEvent::default();
+    /// assert_eq!(event.inner_mut(), &mut std::collections::BTreeMap::default());
+    /// ```
+    #[instrument(level = "trace", skip(self))]
+    pub fn inner_mut(&mut self) -> &mut BTreeMap<String, Value> {
+        &mut self.fields
+    }
+}
+
+impl remap_lang::Object for LogEvent {
+    fn get(&self, path: &remap_lang::Path) -> Result<Option<remap_lang::Value>, String> {
+        let path_string = path.to_string();
+        if path_string == "." {
```

Comment:
We could call `path.is_root` here to see if we are root, which would save the allocation. It gets in the way of the `trace!` below. Could you change the trace to be: `trace!(path = %path.to_string()..` which would then only build in the allocation for debug builds?

Reply:
Heck yes, good idea!